### ISSUE TITLE
Compare “world ready” shaper to generic shaper?

### ISSUE THREAD

**Issue by ctrlcctrlv** [RolandDreger/open-type-features#5]

**Comment by RolandDreger** [comment]
Can you please describe your issue in a little more detail? 😉

**Comment by ctrlcctrlv** [comment]
Yes, InDesign has multiple shapers, maybe your script can compare their results

**Comment by RolandDreger** [comment]
Now I get it. I found a property in InDesign DOM. If I am right it is new since InDesign 2020 and you can change the shaper with it. (Adobe shaping engine vs. HarfBuzz)

I still need to find a proper place and how to test it.

**Comment by ctrlcctrlv** [comment]
Yes, I'm sorry, too big a font nerd hehe. I think it's also changable via UI by the way, so users are likely to come across it. And for many languages it's required, for some of my fonts too even for Latin (as the default shaper doesn't have mult sub support!!)

**Comment by RolandDreger** [comment]
I added an option to change the shaping engine. Tab: Preferences

If you want you can test the script: 

Branch **Develop** or download here   
[openTypeFeatures_v2-1.zip](https://github.com/RolandDreger/open-type-features/files/7988968/openTypeFeatures_v2-1.zip)

>  ## Shaping Engine
> 
> Set Lipika (default) or Harfbuzz as the shaping engine for Latin and Indic script 
> in Adobe World-Ready Composer. The property is applied to the active document. 
> After activation (only) the selected text will be recomposed. Available for InDesign 
> 2020+ only.

Or can you tell me a character combination to test this?

**Comment by ctrlcctrlv** [comment]
@RolandDreger https://github.com/ctrlcctrlv/FRBAmericanCursive/issues/14#issuecomment-986088089 "Performances Report" in any of the FRBAC v1 fonts (v2 will have the same bug so it's immaterial for you to test v1 or v2 for this, the v2 bug fix is just adding docs telling people to enable World-Ready Shaper)

**Comment by RolandDreger** [comment]
I think the problem is that InDesign uses Harfbuzz only in the World Ready Composer. When it is set, Harfbuzz can be used for certain fonts, as I understand it.

However, when I use the Word Ready Composer, the faulty rendering does not occur and change the shaping engine via script has no effect.

<img width="1284" alt="adobe-paragraph-composer" src="https://user-images.githubusercontent.com/19747449/152248876-0cd24f74-e6e4-47c6-a98d-b41e72b5fea3.png">

<img width="1301" alt="world-ready-composer" src="https://user-images.githubusercontent.com/19747449/152248903-48e3f847-01e3-4abb-acbd-f8d67a8fbcb4.png">

**Comment by ctrlcctrlv** [comment]
@RolandDreger I understand the issue's cause quite well.

The normal Adobe iD shaper does not support OpenType GSUB Type 2 (Multiple Substitution).

I require GSUB 2 to do that trick. The lookup is essentially…

```fea
lookup Fina {
    …
    sub s by s.low tail.low;
    …
} Fina;
```

If you don't get a tail, you never get the final transition to `s.low`. My tails are also removable—again via a Mult Sub (with 0 replacement values, fun hack @simoncozens added to fontTools that is supported in CoreText, MS DW and HarfBuzz), that occurs _after_ the addition of the tail, it doesn't work by preventing the tail addition, but rather looks like…

```fea
sub tail.low by NULL;
sub tail.lowwide by NULL;
sub tail.high by NULL;
```

Anyway, my font will never work right with Adobe's broken non-compliant shaper, so I was thinking it'd be a good test for you to have an example of such a font.

**Comment by ctrlcctrlv** [comment]
@RolandDreger Thanks, fixed in e048fcf10583d6f27111d2a504fd31a1a94d1120!

(I'm only adding this comment because otherwise it looks like you closed issue without fixing/adding feature, which is not the case. 🙇‍♂️)
